### PR TITLE
RDOIAT-580 - Allow screenshots for iframes

### PR DESCRIPTION
**Pull request by cangosta**
### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Enhancement (changes that improvement of current feature or performance)
- [ ] Refactoring (changes that neither fixes a bug nor adds a feature)
- [ ] Test Case (changes that add missing tests or correct existing tests)
- [ ] Code style optimization (changes that do not affect the meaning of the code)
- [ ] Docs (changes that only update documentation)
- [ ] Chore (changes that don't modify src or test files)

### Self Check before Merge

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**CONTRIBUTING**](https://github.com/bubkoo/html-to-image/blob/master/CONTRIBUTING.md) document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: src/options.ts [hunk 1]
```diff
@@ -44,6 +44,11 @@ export interface Options {
    * Default: false | undefined - which strips away the query parameters
    */
   includeQueryParams?: boolean
+  /**
```
Comment (cangosta) [inline src/options.ts]: delete this

File: .eslintrc [hunk 2]
```diff
@@ -28,11 +28,14 @@
     "requireConfigFile": false
   },
   "rules": {
+    "no-proto": 0,
```
Comment (cangosta) [inline .eslintrc]: do we still need all these exceptions? I did it for the POC

**Review by tbragaf — APPROVED** [review]

File: src/embedWebFonts.ts [hunk 3]
```diff
@@ -11,56 +11,98 @@ const cssFetchCache: {
...
+): Promise<string> {
   return meta.cssText.then((raw: string) => {
     let cssText = raw
     const regexUrl = /url\(["']?([^"')]+)["']?\)/g
     const fontLocs = cssText.match(/url\([^)]+\)/g) || []
-    const loadFonts = fontLocs.map((location: string) => {
-      let url = location.replace(regexUrl, '$1')
-      if (!url.startsWith('https://')) {
-        url = new URL(url, meta.url).href
+    // const fontLocs5 = [fontLocs[4]]
```
Comment (cangosta) [inline src/embedWebFonts.ts]: delete

File: src/embedWebFonts.ts [hunk 4]
```diff
@@ -11,56 +11,98 @@ const cssFetchCache: {
...
+    const loadFonts = fontLocs.map((loc: string, index: number) => {
+      const url = loc.replace(regexUrl, '$1')
+
+      let urlToFetch = resolveUrl(url, meta.url, document, window)
+
+      if (!urlToFetch.startsWith('https://')) {
+        urlToFetch = new URL(urlToFetch, meta.url).href
       }
 
+      console.warn(` iteration ${index} location: ${urlToFetch}`)
```
Comment (cangosta) [inline src/embedWebFonts.ts]: remove this

File: src/embedWebFonts.ts [hunk 5]
```diff
@@ -11,56 +11,98 @@ const cssFetchCache: {
...
+          console.warn(`Failed to load font ${urlToFetch} `, reason)
+          return Promise.resolve(null)
+        })
+    })
+
+    const timeoutPromise = new Promise<string>((resolve, reject) => {
+      window.setTimeout(() => {
+        console.warn('Timed out')
+        resolve('timeout')
+      }, 5000)
```
Comment (cangosta) [inline src/embedWebFonts.ts]: change to 3s

File: src/embedWebFonts.ts [hunk 6]
```diff
@@ -86,13 +138,12 @@ function parseCSS(source: string) {
     }
     result.push(matches[0])
   }
-  cssText = cssText.replace(keyframesRegex, '')
+  cssText = (cssText as any).replaceAll(keyframesRegex, '')
```
Comment (cangosta) [inline src/embedWebFonts.ts:141]: please check again if we need this change.

File: src/util.ts [hunk 7]
```diff
@@ -23,7 +23,32 @@ export function getMimeType(url: string): string {
...
+  const fetchAbortController = new AbortController()
+  const fetchTimeoutId = setTimeout(() => {
+    fetchAbortController.abort()
+    console.warn(
+      `Fetch exceeded the defined timeout of ${timeoutInMs}ms. Requested url was ${url}`,
+    )
+  }, timeoutInMs)
+  return window
+    .fetch(url, {
+      mode: 'no-cors', // this avoids failures when fetching resources to external domains inside iframes but the response will be empty so the fetch is not successful anyway
```
Comment (cangosta) [inline src/util.ts]: as we talked, it better to remove this.

**Review by cangosta — COMMENTED** [review]
Check comments

File: src/embedWebFonts.ts [hunk 8]
```diff
@@ -11,56 +11,88 @@ const cssFetchCache: {
...
+
+    const timeoutPromise = new Promise<string>((resolve, reject) => {
+      window.setTimeout(() => {
+        resolve('timeout')
+      }, 3000)
     })
 
     // eslint-disable-next-line promise/no-nesting
-    return Promise.all(loadFonts).then(() => cssText)
+    return Promise.race([Promise.all(loadFonts), timeoutPromise]).then(
```
Comment (cangosta) [inline src/embedWebFonts.ts]: replace the usage of race + all to allSettled

**Review by osplm — APPROVED** [review]

**Pull request merged**